From: Russ Allbery <rra@debian.org>
Date: Sun, 22 Mar 2020 21:20:29 -0700
Subject: Fix buffer overflow in prompting

SECURITY: All previous versions of this module could overflow the
buffer provided by the underlying Kerberos library for the response to
a prompt by writing a single nul character past the end of the buffer.
(CVE-2020-10595)
---
 prompting.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/prompting.c b/prompting.c
index 268d459..6492da9 100644
--- a/prompting.c
+++ b/prompting.c
@@ -318,12 +318,13 @@ pamk5_prompter_krb5(krb5_context context UNUSED, void *data, const char *name,
     if (banner != NULL && !args->silent)
         pam_prompts++;
     for (i = 0; i < num_prompts; i++, pam_prompts++) {
-        size_t len;
+        size_t len, allowed;
 
         if (resp[pam_prompts].resp == NULL)
             goto cleanup;
         len = strlen(resp[pam_prompts].resp);
-        if (len > prompts[i].reply->length)
+        allowed = prompts[i].reply->length;
+        if (allowed == 0 || len > allowed - 1)
             goto cleanup;
 
         /*
